deluge/decorators.py

```diff
@@ -159,3 +161,48 @@ def depr_func(*args, **kwargs):
         return func(*args, **kwargs)
 
     return depr_func
+
+
+class CoroutineDeferred(defer.Deferred):
+    """Deferred that will dynamically act like a coroutine instead of a Deferred if it's awaited."""
+
+    def __init__(self, coro):
+        # Delay this import to make sure a reactor was installed first
+        from twisted.internet import reactor
+
+        super().__init__()
+        self.coro = coro
+        self.awaited = None
+        self.activate_deferred = reactor.callLater(0, self.activate)
+
+    def __await__(self):
+        if self.awaited in [None, True]:
+            self.awaited = True
+            return self.coro.__await__()
+        # Already in deferred mode
+        return super().__await__()
+
+    def activate(self):
+        """If the result wasn't awaited before the next context switch, we turn it into a deferred."""
+        if self.awaited is None:
+            self.awaited = False
+            d = defer.Deferred.fromCoroutine(self.coro)
+            d.chainDeferred(self)
+
+    def addCallbacks(self, *args, **kwargs):  # noqa: N802
+        assert not self.awaited, 'Cannot add callbacks to an already awaited coroutine.'
+        self.activate()
+        return super().addCallbacks(*args, **kwargs)
+
+
+def maybe_coroutine(f):
+    """Wraps an async function to make it usable as a normal function that returns a Deferred."""
+
+    @wraps(f)
+    def wrapper(*args, **kwargs):
```

Comment:
Shouldn't this be an async def? Presumably you're using this decorator on async functions, so it should return one?

Reply:
Nope, it turns it into a regular function. That's why it's usable in situations outside of async functions. Works almost the same as the two commented lines below using 'ensureDeferred'.

Reply:
Sorta like this wrapper here. https://stackoverflow.com/a/44627140 `CoroutineDeferred` has _almost_ the same function as `ensureDeferred`, except it will pass the coroutine through directly where appropriate.